### PR TITLE
Generate API documents for Karmada CRD

### LINKED ISSUE #1367: Generates API documents
**What would you like to be added**:
Generates API documents (like [swagger.json](https://github.com/kubernetes/kubernetes/blob/master/api/openapi-spec/swagger.json)).

- [ ] Auto-generates the API documents when API changes.
- [ ] Check If API documents are up to date by CI.

**Why is this needed**:
Simplify API development for users.

### PR DESCRIPTION
**Pull request by lonelyCZ**
Signed-off-by: lonelyCZ <531187475@qq.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Automatically generate API document for Karmada CRD.

**Which issue(s) this PR fixes**:
Fixes #1367 

**Special notes for your reviewer**:
I researched this function in multiple project these days, and I found that most projects implement CRUD of the APIs  by themselves and then regenerate it into documentation.

If we only want the default CRUD interface, we need to register the CRD into the API Server first and then export the default interface.

There is no better way yet and what do you think of this？

/cc @RainbowMango 

Now just to show the effect, there's a lot of detail need to be completed.

image

**Does this PR introduce a user-facing change?**:

```release-note
Introduced Swagger API docs.
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Looks pretty good.
cc me after ready for review.

**Comment by lonelyCZ** [comment]
This pr can be reviewed~

I have two uncertain point: 
1. I directly use `gclient` to register all schemes, 
https://github.com/karmada-io/karmada/blob/e542c347af2cd151f71a6406d812b985c01a6ca9/hack/tools/swagger/generateswagger.go#L28

2. Generating all `zz_openapi` to one file
https://github.com/karmada-io/karmada/blob/e542c347af2cd151f71a6406d812b985c01a6ca9/hack/update-codegen.sh#L128-L142

And this function refer to https://github.com/kubesphere/kubesphere/tree/master/api/openapi-spec

Looking forward to your review. :) @RainbowMango

**Comment by RainbowMango** [comment]
> And this function refer to https://github.com/kubesphere/kubesphere/tree/master/api/openapi-spec

What do you mean? Do you get the idea from the `kubesphere` project?

**Comment by RainbowMango** [comment]
This PR does two things:
- Generates the open API (zz_generated.openapi.go)
- Then, generates the swagger API documents(swagger.json)

Right?

File: pkg/apis/work/v1alpha1/binding_types.go [hunk 1]
```diff
@@ -6,6 +6,26 @@ import (
...
+	// ResourceKindClusterResourceBinding is kind name of ClusterResourceBinding
+	ResourceKindClusterResourceBinding = "ClusterResourceBinding"
+	// ResourceSingularClusterResourceBinding is singular name of ClusterResourceBinding
+	ResourceSingularClusterResourceBinding = "clusterresourcebinding"
+	// ResourcePluralClusterResourceBinding is kind plural of ClusterResourceBinding
+	ResourcePluralClusterResourceBinding = "clusterresourcebindings"
+	// ResourceNamespaceScopedClusterResourceBinding indicates if ClusterResourceBinding is NamespaceScoped
+	ResourceNamespaceScopedClusterResourceBinding = false
+)
+
```
Comment (RainbowMango) [inline pkg/apis/work/v1alpha1/binding_types.go:28]: The `v1alpha1` version has been deprecated, do we still have to generate the API for it?

File: hack/tools/swagger/generateswagger.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	mapper.AddSpecific(workv1alpha1.SchemeGroupVersion.WithKind(workv1alpha1.ResourceKindClusterResourceBinding),
+		workv1alpha1.SchemeGroupVersion.WithResource(workv1alpha1.ResourcePluralClusterResourceBinding),
+		workv1alpha1.SchemeGroupVersion.WithResource(workv1alpha1.ResourceSingularClusterResourceBinding), meta.RESTScopeRoot)
+
+	mapper.AddSpecific(workv1alpha1.SchemeGroupVersion.WithKind(workv1alpha1.ResourceKindWork),
+		workv1alpha1.SchemeGroupVersion.WithResource(workv1alpha1.ResourcePluralWork),
+		workv1alpha1.SchemeGroupVersion.WithResource(workv1alpha1.ResourceSingularWork), meta.RESTScopeRoot)
+
+	mapper.AddSpecific(workv1alpha2.SchemeGroupVersion.WithKind(workv1alpha1.ResourceKindResourceBinding),
+		workv1alpha2.SchemeGroupVersion.WithResource(workv1alpha1.ResourcePluralResourceBinding),
```
Comment (RainbowMango) [inline hack/tools/swagger/generateswagger.go:85]: `workv1alpha2`?

File: hack/generate-swagger.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${REPO_ROOT}"/hack/util.sh
+
+util::verify_go_version
+
+go run hack/tools/swagger/generateswagger.go VERSION > api/openapi-spec/swagger.json
```
Comment (RainbowMango) [inline hack/generate-swagger.sh:12]: end of line.

File: hack/tools/swagger/generateswagger.go (hunk 2 above)
Comment (lonelyCZ) [reply hack/tools/swagger/generateswagger.go:85]: Sorry, will fix.

**Comment by lonelyCZ** [comment]
> OK. I'll look at the blog. By the way, any other projects have you looked at?

I also found there is 'api/openapi/swagger.json' at projects: [argo-workflow](https://github.com/argoproj/argo-workflows), [kubevirt](https://github.com/kubevirt/kubevirt).

File: hack/tools/swagger/generateswagger.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	networkingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/networking/v1alpha1"
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha1 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	generatedopenapi "github.com/karmada-io/karmada/pkg/generated/openapi"
+	"github.com/karmada-io/karmada/pkg/util/gclient"
+)
+
+func main() {
+	Scheme := gclient.NewSchema()
```
Comment (lonelyCZ) [inline hack/tools/swagger/generateswagger.go:24]: Due to 'pkg/apis/cluster/v1alpha1' has been implement `ClusterProxyOptions{}`, so need not to use `pkg/apis/cluster` package.

File: hack/generate-swagger.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+set -o pipefail
+
+REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+source "${REPO_ROOT}"/hack/util.sh
+
+util::verify_go_version
+
+go run hack/tools/swagger/generateswagger.go > api/openapi-spec/swagger.json
+
+# Delete trash of generating swagger doc
```
Comment (lonelyCZ) [inline hack/generate-swagger.sh:14]: `defer os.Remove()` in `render.go` can't delete content of `/tmp`, so remove it here.

File: hack/tools/swagger/lib/render.go [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+			}
+			storage[info.gvk.Version][r.Resource] = &StandardREST{info}
+			// Add status router for all resources.
+			storage[info.gvk.Version][r.Resource+"/status"] = &StatusREST{StatusInfo{
+				gvk: info.gvk,
+				obj: info.obj,
+			}}
+
+			// To define additional endpoints for CRD resources, we need to
+			// implement our own REST interface and add it to our custom path.
```
Comment (lonelyCZ) [inline hack/tools/swagger/lib/render.go:100]: Put the comment of customizing path in here?

File: hack/generate-swagger.sh [hunk 7]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
```
Comment (RainbowMango) [inline hack/generate-swagger.sh:1]: Can we update this file name to `hack/update-swagger-docs.sh`? 
to stay in step with other scripts started with `hack/update-xxx.sh`.

**Review by RainbowMango — COMMENTED** [review]
Just one request left.

File: hack/generate-swagger.sh (hunk 7 above)
Comment (lonelyCZ) [reply hack/generate-swagger.sh:1]: Ok.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

Good job!!! @lonelyCZ

**Pull request merged**